PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
-        shex.nShow = static_cast<WORD>(cmdShow);
-        
-
-        Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
-        if (!ShellExecuteEx(&shex))
+        ErrorInformation endingScriptError = RunScript(endScriptInformation, packageRoot, dirStr, cmdShow);
+        error.AddExeName(L"Powershell.exe");
+
+        //If there is an existing error from Monitor or the packaged exe
+        if (error.IsThereAnError())
         {
-            std::wostringstream ss;
-            auto err{ ::GetLastError() };
-            // Remove the ".\r\n" that gets added to all messages
-            auto msg = widen(std::system_category().message(err));
-            msg.resize(msg.length() - 3);
-            ss << L"ERROR: Failed to create detoured shell process\n  Path: \"" << exeName << "\"\n  Error: " << msg << " (" << err << ")";
-            ::PSFReportError(ss.str().c_str());
-            return err;
+            ::PSFReportError(error.Print().c_str());
+            return error.GetErrorNumber();
         }
-        else
-        { 
-            // Propagate exit code to caller, in case they care
-            switch (::WaitForSingleObject(shex.hProcess, INFINITE))
-            {
-            case WAIT_OBJECT_0: // SUCCESS
-                if (((unsigned long long)(shex.hInstApp)) > 32)
-                    return 0;
-                else
-                    return (int)(unsigned long long)(shex.hInstApp);
-            case WAIT_FAILED:
-                return ::GetLastError();
-
-            default:
-                return ERROR_INVALID_HANDLE;
-            }			
+        else if (endingScriptError.IsThereAnError())
+        {
+            ::PSFReportError(endingScriptError.Print().c_str());
+            return endingScriptError.GetErrorNumber();
         }
     }
+
+    return 0;
 }
 catch (...)
 {
     ::PSFReportError(widen(message_from_caught_exception()).c_str());
     return win32_from_caught_exception();
+}
 
+ErrorInformation RunScript(const psf::json_object * scriptInformation, std::filesystem::path packageRoot, LPCWSTR dirStr, int cmdShow)
+{
+    //Script arguments are optional.
+    auto scriptArgumentsJObject = scriptInformation->try_get("scriptArguments");
+    std::wstring scriptArguments;
+    if (scriptArgumentsJObject)
+    {
+        scriptArguments = scriptArgumentsJObject->as_string().wide();
+    }
+
+    std::wstring scriptPath = scriptInformation->get("scriptPath").as_string().wide();
+    std::filesystem::path powershellScriptPath(scriptPath);
+
+    std::wstring powershellCommandString(L"powershell.exe -file ");
+
+    powershellCommandString.append(scriptPath);
+    powershellCommandString.append(L" ");
+    powershellCommandString.append(scriptArguments);
+
+    Log("Looking for the script here: ");
+    LogStringW("Script Path", scriptPath.c_str());
+    auto currentDirectory = (packageRoot / dirStr);
+    auto doesFileExist = std::filesystem::exists(currentDirectory / powershellScriptPath);

```

Comment:
nit: not needed

Reply:
Right.  I don't huh.  Because if CreateProcess fails because it can't find the Script PSFReportError throws up a message box with the message.